### PR TITLE
T-4 Add main data layer

### PR DESCRIPTION
**Pull request by DashaDrahun**
Created Domain Data Entities, Added Migrations, Changed ApplicationUser on AppUser(DAL entity) in ASP Identity classes

### CODE REVIEW

File: PortfolioKeeper/DAL/Context/PortfolioKeeperDbContext.cs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+                .WillCascadeOnDelete(false);
+
+            modelBuilder.Entity<MetallType>()
+                .HasMany(e => e.IngotPrices)
+                .WithRequired(e => e.MetallType)
+                .WillCascadeOnDelete(false);
+
+            modelBuilder.Entity<ExchangeRate>()
+                .Property(e => e.SellRate)
+                .HasPrecision(18, 0);
```
Comment (AntonBessarabov) [inline PortfolioKeeper/DAL/Context/PortfolioKeeperDbContext.cs]: Change to .HasPrecision(9, 4) please

File: PortfolioKeeper/DAL/Context/PortfolioKeeperDbContext.cs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+            modelBuilder.Entity<ExchangeRate>()
+                .Property(e => e.BuyRate)
+                .HasPrecision(18, 0);
```
Comment (AntonBessarabov) [inline PortfolioKeeper/DAL/Context/PortfolioKeeperDbContext.cs]: Change to .HasPrecision(9, 4) please

File: PortfolioKeeper/DAL/Context/PortfolioKeeperDbContext.cs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                .HasMany(e => e.Assets)
+                .WithRequired(e => e.Portfolio)
+                .WillCascadeOnDelete(false);
+
+            modelBuilder.Entity<Portfolio>()
+                .HasMany(e => e.PortfolioValuations)
+                .WithRequired(e => e.Portfolio)
+                .WillCascadeOnDelete(false);
+
+            modelBuilder.Entity<PortfolioValuation>()
```
Comment (AntonBessarabov) [inline PortfolioKeeper/DAL/Context/PortfolioKeeperDbContext.cs]: This is the default value (line 106)

File: PortfolioKeeper/DAL/Context/PortfolioKeeperDbContext.cs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                .WithRequired(e => e.AssetType)
+                .WillCascadeOnDelete(false);
+
+            modelBuilder.Entity<AssetPropertyType>()
+                .HasMany(e => e.AssetPropertyValues)
+                .WithRequired(e => e.AssetTypeProperty)
+                .HasForeignKey(e => e.PropertyTypeId)
+                .WillCascadeOnDelete(false);
+
+            modelBuilder.Entity<AssetValuation>()
```
Comment (AntonBessarabov) [inline PortfolioKeeper/DAL/Context/PortfolioKeeperDbContext.cs]: This is default value (line 69)

File: PortfolioKeeper/DAL/Entities/IngotPrice.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+        public byte CurrencyId { get; set; }
+
+        public byte BankId { get; set; }
+
+        [Column(TypeName = "money")]
+        public decimal SellPrice { get; set; }
+
+        [Column(TypeName = "money")]
+        public decimal BuyPrice { get; set; }
```
Comment (AntonBessarabov) [inline PortfolioKeeper/DAL/Entities/IngotPrice.cs]: I think this field should be nullable

File: PortfolioKeeper/DAL/Entities/IngotPrice.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        public int Id { get; set; }
+
+        public byte MetallTypeId { get; set; }
+
+        public byte CurrencyId { get; set; }
+
+        public byte BankId { get; set; }
+
+        [Column(TypeName = "money")]
+        public decimal SellPrice { get; set; }
```
Comment (AntonBessarabov) [inline PortfolioKeeper/DAL/Entities/IngotPrice.cs]: I think this field should be nullable

File: PortfolioKeeper/DAL/Context/PortfolioKeeperDbContext.cs (hunk 1 above)
Comment (DashaDrahun) [reply PortfolioKeeper/DAL/Context/PortfolioKeeperDbContext.cs]: Done

File: PortfolioKeeper/DAL/Context/PortfolioKeeperDbContext.cs (hunk 2 above)
Comment (DashaDrahun) [reply PortfolioKeeper/DAL/Context/PortfolioKeeperDbContext.cs]: Done

File: PortfolioKeeper/DAL/Entities/IngotPrice.cs (hunk 5 above)
Comment (DashaDrahun) [reply PortfolioKeeper/DAL/Entities/IngotPrice.cs]: Done

File: PortfolioKeeper/DAL/Entities/IngotPrice.cs (hunk 6 above)
Comment (DashaDrahun) [reply PortfolioKeeper/DAL/Entities/IngotPrice.cs]: Done

**Comment by AntonBessarabov** [comment]
Dal entities looks good for me

File: PortfolioKeeper/DAL/Entities/AssetPropertyValue.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+        [Column(TypeName = "xml")]
+        [Required]
+        public string Value { get; set; }
+
+        public DateTime UpdatedOn { get; set; }
+
+        public virtual Asset Asset { get; set; }
+
+        public virtual AssetPropertyType AssetTypeProperty { get; set; }
```
Comment (stas-litvinenka) [inline PortfolioKeeper/DAL/Entities/AssetPropertyValue.cs]: AssetTypeProperty -> AssetPropertyType

File: PortfolioKeeper/DAL/Entities/AssetType.cs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+        [Required]
+        [StringLength(50)]
+        public string Name { get; set; }
+
+        public DateTime CreatedOn { get; set; }
+
+        public virtual ICollection<Asset> Assets { get; set; }
+        
+        public virtual ICollection<AssetType> ChildrenTypes { get; set; }
```
Comment (stas-litvinenka) [inline PortfolioKeeper/DAL/Entities/AssetType.cs]: No need in this property

File: PortfolioKeeper/DAL/Entities/AssetType.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+        public virtual AssetType ParentType { get; set; }
+
+        public virtual ICollection<AssetPropertyType> AssetTypeProperties { get; set; }
```
Comment (stas-litvinenka) [inline PortfolioKeeper/DAL/Entities/AssetType.cs]: AssetTypeProperties -> AssetPropertieTypes

File: PortfolioKeeper/DAL/Context/PortfolioKeeperDbContext.cs [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        public virtual DbSet<Asset> Assets { get; set; }
+        public virtual DbSet<AssetPropertyValue> AssetPropertyValues { get; set; }
+        public virtual DbSet<AssetType> AssetTypes { get; set; }
+        public virtual DbSet<AssetPropertyType> AssetTypeProperties { get; set; }
+        public virtual DbSet<AssetValuation> AssetValuations { get; set; }
+        public virtual DbSet<Bank> Banks { get; set; }
+        public virtual DbSet<ExchangeRate> ExchangeRates { get; set; }
+        public virtual DbSet<Portfolio> Portfolios { get; set; }
+        public virtual DbSet<PortfolioValuation> PortfolioValuations { get; set; }
+        public virtual DbSet<Сurrency> Сurrency { get; set; }
```
[the remark this answers is not in the document]
Comment (DashaDrahun) [reply PortfolioKeeper/DAL/Context/PortfolioKeeperDbContext.cs]: Done

File: PortfolioKeeper/DAL/Entities/AssetPropertyValue.cs (hunk 7 above)
Comment (DashaDrahun) [reply PortfolioKeeper/DAL/Entities/AssetPropertyValue.cs]: Done

File: PortfolioKeeper/Data/Context/PortfolioKeeperDbContext.cs [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+
+            modelBuilder.Entity<ExchangeRate>()
+                .Property(e => e.SellRate)
+                .HasPrecision(9, 4);
+
+            modelBuilder.Entity<ExchangeRate>()
+                .Property(e => e.BuyRate)
+                .HasPrecision(9, 4);
+
+            //modelBuilder.Entity<Asset>()
```
Comment (stas-litvinenka) [inline PortfolioKeeper/Data/Context/PortfolioKeeperDbContext.cs]: Please remove commented code.

**Comment by stas-litvinenka** [comment]
[T-4 Add main data layer](https://trello.com/c/cW8wINbm/3-t-4-add-main-data-layer)

**Pull request merged**